### ISSUE TITLE
#[derive(Debug)] is triggering the unused_results lint

### ISSUE THREAD

**Issue by canndrew** [rust-lang/rust#29710]
The following code fails to build on nightly

```rust
#![deny(unused_results)]

#[derive(Debug)]
enum Foo {
    Bar(u32),
}
```
It generates this error:
```
<anon>:3:10: 3:15 error: unused result
<anon>:3 #[derive(Debug)]
                  ^~~~~
<anon>:3:10: 3:15 note: in this expansion of #[derive_Debug] (defined in <anon>)
<anon>:1:9: 1:23 note: lint level defined here
<anon>:1 #![deny(unused_results)]
```

**Comment by apasel422** [comment]
The expansion is:

```rust
#[automatically_derived]
#[allow(unused_qualifications)]
impl ::std::fmt::Debug for Foo {
    fn fmt(&self, __arg_0: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
        match (&*self,) {
            (&Foo::Bar(ref __self_0),) => {
                let mut builder = __arg_0.debug_tuple("Bar");
                builder.field(&&(*__self_0));
                builder.finish()
            }
        }
    }
}
```
The call to `builder.field(...)` returns `self` for chaining purposes, but the expansion obviously ignores it.

**Comment by apasel422** [comment]
This was a result of #29565.
